### PR TITLE
Vscode fixes

### PR DESCRIPTION
**Pull request by roironn**

### CODE REVIEW

File: packages/vscode/src/salto/context.ts [hunk 1]
```diff
@@ -76,22 +76,23 @@ const getText = (content: string, range: EditorRange): string => {
...
-    const path = contextRange.name.slice(element.elemID.getFullName().length + 1)
-    return { element, path, isList }
+  const rangeContent = getText(fileContent, contextRange.range)
+  const isList = _.last(rangeContent) === ']' || _.takeRight(rangeContent, 2).join('') === '],'
+  const elemID = ElemID.fromFullName(contextRange.name)
+  const { parent, path } = elemID.createTopLevelParentID()
+  const topLevelParent = refElements[parent.getFullName()]
+  if (elemID.idType === 'field') {
+    const fieldName = elemID.getFullNameParts()[3]
+    const fieldPath = elemID.getFullNameParts().slice(4)
```
Comment (roironn) [inline packages/vscode/src/salto/context.ts:89]: Shouldn't be here

File: packages/vscode/src/salto/symbols.ts [hunk 2]
```diff
@@ -31,34 +32,28 @@ export interface SaltoSymbol {
...
-    const fullName = context.ref.path
-      ? context.ref.path
-      : context.ref.element.elemID.name
-    if (prefName && fullName.indexOf(prefName) >= 0) {
-      const partName = fullName.slice(fullName.indexOf(prefName) + prefName.length + 1)
-      return Number.isNaN(Number(partName)) ? partName : `[${partName}]`
-    }
-    return fullName
+    const fullName = _.last(context.ref.path) || context.ref.element.elemID.name
+    return Number.isNaN(Number(fullName)) ? fullName : `[${fullName}]`
```
Comment (roironn) [inline packages/vscode/src/salto/symbols.ts:38]: use isIndexPart

File: packages/adapter-api/src/utils.ts [hunk 3]
```diff
@@ -34,60 +35,84 @@ export const getDeepInnerType = (listType: ListType): ObjectType | PrimitiveType
...
+    if (isIndexPathPart(key) && isListType(source)) {
+      return source.innerType
+    }
     if (source.annotationTypes[key]) return source.annotationTypes[key]
     if (isObjectType(source)) return source.fields[key]
     return undefined
   }
 
-  const [curPart, ...restOfParts] = pathParts.filter(p => Number.isNaN(Number(p)))
+  const [curPart, ...restOfParts] = pathParts// .filter(p => Number.isNaN(Number(p)))
```
Comment (roironn) [inline packages/adapter-api/src/utils.ts:51]: Avshalom Core - remove comment

File: packages/adapter-api/src/utils.ts [hunk 4]
```diff
@@ -34,60 +35,84 @@ export const getDeepInnerType = (listType: ListType): ObjectType | PrimitiveType
...
-  // and we can't tell of the _ is path separator or a part of the
-  // the path name. As long as path is not empty we will try to advance
-  // in the recursion in two ways - First we try only the first token.
-  // If it fails, we try to first to tokens (the recursion will take)
-  // care of the next "join"
-
-  // We start by filtering out numbers from the path as they are
-  // list indexes, which are irrelevant for type extractions
+export const getSubElement = (baseType: TypeElement, pathParts: string[]): {
+  field?: Field
```
Comment (roironn) [inline packages/adapter-api/src/utils.ts:39]: extract to type

File: packages/adapter-api/src/utils.ts [hunk 5]
```diff
@@ -34,60 +35,84 @@ export const getDeepInnerType = (listType: ListType): ObjectType | PrimitiveType
...
+    if (_.isEmpty(curPart)) {
+      return fieldType
+    }
+    if (isIndexPathPart(curPart) && isListType(fieldType)) {
+      return getFieldInternalType(fieldType.innerType, restOfParts)
+    }
+    return undefined
+  }
+  const fieldData = getField(baseType, path)
+  const r = fieldData?.field && getFieldInternalType(fieldData.field.type, fieldData.path)
```
Comment (roironn) [inline packages/adapter-api/src/utils.ts:105]: debug leftover

File: packages/vscode/src/salto/completions/suggestions.ts [hunk 6]
```diff
@@ -197,8 +197,11 @@ export const fieldSuggestions = (params: SuggestionsParams): Suggestions => {
 export const fieldValueSuggestions = (params: SuggestionsParams): Suggestions => {
   if (!(params.ref && isInstanceElement(params.ref.element))) return []
   const attrName = params.tokens[0]
-  const valueField = getField(params.ref.element.type, params.ref.path)?.field
-  const valueFieldType = getFieldType(params.ref.element.type, params.ref.path)
+  const refPathWithoutAttr = _.last(params.ref.path) !== attrName && !params.ref.isList
```
Comment (roironn) [inline packages/vscode/src/salto/completions/suggestions.ts:203]: [AC] should be WITH not without

File: packages/vscode/src/salto/definitions.ts [hunk 7]
```diff
@@ -31,8 +30,11 @@ export const provideWorkspaceDefinition = async (
...
-    return getLocations(workspace, context.ref.element.elemID.getFullName())
   }
-  // We are not in instance, so we can just look the current token
-  return getLocations(workspace, parseElemID(token).getFullName())
+  // We are not in instance field, so we can just look the current token
+  try {
+    return getLocations(workspace, token)
+  } catch (e) {
+    return []
+  }
```
Comment (roironn) [inline packages/vscode/src/salto/definitions.ts:39]: Test this

File: packages/vscode/src/salto/usage.ts [hunk 8]
```diff
@@ -26,33 +28,48 @@ const getUsages = async (
...
-): Promise<SaltoElemLocation[]> => ([
-  ..._.flatten(await Promise.all(
-    (await workspace.elements).map(e => getUsages(workspace, e, token))
-  )),
-  ...await getLocations(workspace, token),
-])
+  token: string,
+  context: PositionContext
+): Promise<SaltoElemLocation[]> => {
+  const fullName = context.ref?.element.elemID.getFullName() ?? token
```
Comment (roironn) [inline packages/vscode/src/salto/usage.ts:69]: add test for token branch or remove token branch

File: packages/vscode/src/commands.ts [hunk 9]
```diff
@@ -30,13 +30,12 @@ export const createCopyReferenceCommand = (
   }
   const position = editor.selection.active
   await workspace.awaitAllUpdates()
-  const validWorkspace = await workspace.getValidCopy()
-  if (!validWorkspace) {
+  if (!workspace) {
```
Comment (roironn) [inline packages/vscode/src/commands.ts:33]: not needed (for all)

File: packages/vscode/src/events.ts [hunk 10]
```diff
@@ -21,17 +21,23 @@ import { EditorWorkspace } from './salto/workspace'
 import { getDiagnostics } from './salto/diagnostics'
 import { toVSDiagnostics } from './adapters'
 
+const DIAG_IDLE_PERIOD = 500
+let diagTimout: NodeJS.Timeout
```
Comment (roironn) [inline packages/vscode/src/events.ts:25]: use debounce instead of this

File: packages/vscode/test/salto/workspace.test.ts [hunk 11]
```diff
@@ -65,26 +65,26 @@ describe('workspace', () => {
...
-    const shouldBeCurrent = await workspace.getValidCopy()
-    if (!shouldBeCurrent) throw new Error('lastValid not defined')
-    expect([...(await shouldBeCurrent.errors()).all()])
-      .toEqual([...(await workspace.errors()).all()])
-    expect(await shouldBeCurrent.elements).toEqual(await workspace.elements)
+  // it('should return last valid state if there are errors', async () => {
+  //   const baseWs = await mockWorkspace(naclFileName)
+  //   baseWs.hasErrors = jest.fn().mockResolvedValue(false)
+  //   const workspace = new EditorWorkspace(workspaceBaseDir, baseWs)
+  //   const shouldBeCurrent = await workspace.getValidCopy()
```
Comment (roironn) [inline packages/vscode/test/salto/workspace.test.ts:72]: delete

File: packages/vscode/src/salto/completions/suggestions.ts [hunk 12]
```diff
@@ -272,6 +275,10 @@ export const typesSuggestions = (params: SuggestionsParams): Suggestions => {
   ]
 }
 
+export const typeBodySuggestions = (params: SuggestionsParams): Suggestions => [
```
Comment (roironn) [inline packages/vscode/src/salto/completions/suggestions.ts:282]: tests

File: packages/vscode/src/salto/completions/provider.ts [hunk 13]
```diff
@@ -83,14 +83,21 @@ const getLineType = (
   if (context.type === 'type' && isDefLine) {
     return 'type'
   }
-  if (context.type === 'type' && !isDefLine) {
+  if (context.type === 'type'
```
Comment (roironn) [inline packages/vscode/src/salto/completions/provider.ts:86]: squash this commit with the previous commit

Comment (roironn) [reply packages/vscode/src/salto/completions/provider.ts:86]: 30074b971e642d89ef8b8a54323ff1c68c31e780

File: packages/vscode/src/salto/completions/suggestions.ts [hunk 14]
```diff
@@ -236,17 +236,21 @@ export const annoSuggestions = (params: SuggestionsParams): Suggestions => {
 
 export const annoValueSuggestions = (params: SuggestionsParams): Suggestions => {
   if (!params.ref) return []
-  const annoName = params.tokens[0]
+
+  const attrName = params.tokens[0]
+  const refPathWithAttr = _.last(params.ref.path) !== attrName && !params.ref.isList
```
Comment (roironn) [inline packages/vscode/src/salto/completions/suggestions.ts:241]: extract to function

File: packages/vscode/src/events.ts (hunk 10 above)
Comment (erikzaadi) [reply packages/vscode/src/events.ts:25]: Deferred?

File: packages/vscode/src/providers.ts [hunk 15]
```diff
@@ -55,17 +55,16 @@ export const createCompletionsProvider = (
     position: vscode.Position
   ) => {
     await workspace.awaitAllUpdates()
-    const validWorkspace = await workspace.getValidCopy()
-    if (validWorkspace) {
+    if (workspace) {
```
Comment (erikzaadi) [inline packages/vscode/src/providers.ts:58]: [AC] Redundant

File: packages/vscode/src/providers.ts [hunk 16]
```diff
@@ -79,15 +78,14 @@ export const createDefinitionsProvider = (
     doc: vscode.TextDocument,
     position: vscode.Position,
   ): Promise<vscode.Definition> => {
-    const validWorkspace = await workspace.getValidCopy()
-    if (validWorkspace) {
+    if (workspace) {
```
Comment (erikzaadi) [inline packages/vscode/src/providers.ts:81]: [AC].push

File: packages/vscode/src/salto/completions/suggestions.ts (hunk 14 above)
Comment (erikzaadi) [reply packages/vscode/src/salto/completions/suggestions.ts:241]: Was this not needed in the end?

File: packages/vscode/test/salto/workspace.test.ts (hunk 11 above)
Comment (erikzaadi) [reply packages/vscode/test/salto/workspace.test.ts:72]: Resolved?

File: packages/vscode/test/salto/workspace.test.ts [hunk 17]
```diff
@@ -65,26 +65,26 @@ describe('workspace', () => {
...
-    workspace.setNaclFiles({ filename: 'error', buffer: 'error' })
-    await workspace.awaitAllUpdates()
-    expect(await workspace.elements).toBeDefined()
-    expect(workspace.hasErrors()).toBeTruthy()
-    await validate(workspace, 17)
-    const lastValid = workspace.getValidCopy()
-    if (!lastValid) throw new Error('lastValid not defined')
-    expect(lastValid).not.toEqual(workspace)
-  })
+  //   baseWs.hasErrors = jest.fn().mockResolvedValue(true)
```
Comment (erikzaadi) [inline packages/vscode/test/salto/workspace.test.ts:78]: DEM SLASHEZ

File: packages/vscode/src/events.ts [hunk 18]
```diff
@@ -17,22 +17,28 @@ import * as vscode from 'vscode'
...
-): Promise<void> => {
-  await workspace.awaitAllUpdates()
-  const newDiag = toVSDiagnostics(
-    workspace.baseDir,
-    await getDiagnostics(workspace)
-  )
-  diagCollection.set(newDiag)
-}
+): (
+) => void => _.debounce(
```
Comment (erikzaadi) [inline packages/vscode/src/events.ts:30]: 😻

File: packages/vscode/src/providers.ts (hunk 15 above)
Comment (erikzaadi) [reply packages/vscode/src/providers.ts:58]: STILL **[AC]**

File: packages/vscode/src/providers.ts (hunk 16 above)
Comment (erikzaadi) [reply packages/vscode/src/providers.ts:81]: STILL **[AC]**

File: packages/vscode/src/salto/completions/suggestions.ts [hunk 19]
```diff
@@ -42,6 +42,12 @@ export const isInsertText = (value: any): value is InsertText => (
   value.label !== undefined && value.insertText !== undefined
 )
 
+const getRefPathWithAttr = (attrName: string, ref: ContextReference): string[] => (
```
Comment (erikzaadi) [inline packages/vscode/src/salto/completions/suggestions.ts:45]: 😻

File: packages/vscode/src/salto/usage.ts [hunk 20]
```diff
@@ -26,33 +28,49 @@ const getUsages = async (
...
-    const typeDumpName = dumpElemID(element.type)
-    return (typeDumpName === fullName)
-      ? getLocations(workspace, element.elemID.getFullName())
-      : []
+  if (isInstanceElement(element) && element.type.elemID.getFullName() === fullName) {
+    pathesToAdd.add(element.elemID)
   }
-  return []
+  const transformFunc = ({ value, field, path }: TransformFuncArgs): Value => {
+    if (field?.elemID.getFullName() === fullName && path && !isIndexPathPart(path.name)) {
```
Comment (erikzaadi) [inline packages/vscode/src/salto/usage.ts:45]: Wanna add a comment about the field value here?

**Pull request merged**